Rewriting history (#19)
ImperialCollegeLondon/intermediate_grad_school_git_course: _episodes/l1-03-intermediate_concepts.md

objectives:
- "Understand the options for rewriting git history"
- "Know how to use them effectively when working with collaborators"
- "Understand the risks associated with rewriting history"
keypoints:
- "There are several ways of rewriting git history, each with specific use cases associated to them"
- "Rewriting history can have unexpected consequences and you risk loosing information irremedaibly"
- "Reset: You have made a mistake and want to keep the commit history tidy for the benefit of collaborators"
- "Revert: You want to undo something done in the past without messing too much with the timeline, upsetting your collaborators"
- "Stash: You want to do somnething else - eg. checkout someone else's branch - without loosing your current work"
- "Rebase: Someone else has updated the main branch while you've been working and need to bring those changes to your branch"
---

## Rewriting history with Git

While version control is useful to keep track of changes made to a piece of work over
time, it also let you to modify the timeline of commits. There are several totally
ligitimate reasons why you might want to do that, from keeping the commit history clean

> of unsuccessful attempts to do something to incorporate work done by someone else.

Reviewer: Good idea. Could write it in a user story-type way, e.g. "Alice has done X and wants to do Y..."

> of unsuccessful attempts to do something to incorporate work done by someone else.

Author: I'm not sure I follow... Could you clarify what you mean?

> of unsuccessful attempts to do something to incorporate work done by someone else.

Author: I'm genuinely unsure about what you mean by tangle examples or case studies. Do you have any specific - complete - suggestion of what one of these might look like?

> of unsuccessful attempts to do something to incorporate work done by someone else.

Reviewer: I was thinking of just one-line descriptions of the sorts of scenarios these techniques can help with, e.g. just add a paragraph like so:

```
For example:
- You have already made a commit, but realise there are other changes you forgot to include
- You have made a commit, but have changed your mind and want to remove this change from your history
- You want to undo a previous commit made by yourself or a colleague (and make a record of this in the history)
- You want to "move" one or more commits so they are based on top of some other work (e.g. new changes made to the main branch
```

> > |/
> > * 57d4505 Revert "Added instruction to enjoy"
> > * 5cb4883 Added 1/2 onion
> > * 43536f3 Added instruction to enjoy
> > * 745fb8b Adding ingredients and instructions
> > ~~~
> > {: .output}
> > Note that while the `experiment` branch still mentions the adjustment of salt, that is
> > no longer part of the `main` commit history. Your working directory has become identical
> > to before starting the salty that adventure.
> >
> {: .solution}
>
{: .challenge}
> ## Changing History Can Have Unexpected Consequences
>
> Using `git reset` to remove a commit is a bad idea if
> you have **shared** it yet with other people. If you make a commit and share it on
> GitHub or with a colleague by other means then removing that commit from your Git
> history will cause inconsistencies that may be difficult to resolve later. We
> only recommend this approach for commits that are only in your local working
> copy of a repository.
{: .callout}

### Removing branches once you are done with them is good practice
Over time, you will accumulate lots of branches to implement different features in you
code. It is good practice to remove them once they have fulfil their purpose. You can do

> that using the `-D` flag with the `git branch` command:

Author: At this point in the course, I would keep things simple and settle with `-D`. It is drastic, but simple to consider.

> >
> > ## Solution
> > ~~~
> > $ git branch -D experiment
> > $ git graph
> > ~~~
> > {: .commands}
> >
> > Now, the commit history should look as:
> > ~~~
> > *   fe0d257 (HEAD -> main) Merge branch 'experiment'
> > |\
> > | * 99b2352 Reduced the amount of coriander
> > * | 2c2d0e2 Merge branch 'experiment'
> > |\|
> > | * d9043d2 Try with some coriander
> > * | 6a2a76f Corrected typo in ingredients.md
> > |/
> > * 57d4505 Revert "Added instruction to enjoy"
> > * 5cb4883 Added 1/2 onion
> > * 43536f3 Added instruction to enjoy
> > * 745fb8b Adding ingredients and instructions
> > ~~~
> > {: .output}
> >
> > Now there is truly no trace of your attempts to change the content of salt!
> >
> {: .solution}
>
{: .challenge}
### Reversing a commit

As pointed out, using `reset` can be dangerous and it is not suitable if you need to be
more surgical in what you want to change, affecting just what was done on a commit a
while ago, potentially already shared with collaborators. To address that, we have
`revert`.

`git revert` creates a commit that exactly cancels the changes made by a previous one.
It is a new commit, so it is part of the history, but its purpose is to undo something

> done in the past.

Author: It is explained just in the previous paragraph. Isn't it enough?

> done in the past.

Reviewer: On second glace, actually that does look fine. Sorry, I think I must have skipped over that paragraph!

The syntax in this case is:
~~~

> $ git revert --no-edit COMMIT_HASH

Author: I think I can mention that they can omit the `--no-edit` and add a message, but I rather leave it there as default so there is, at least, a message saying that certain other commit was reverted.

> $ git revert --no-edit COMMIT_HASH

Reviewer: If you omit `--no-edit` your text editor will open up with the default message about reverting commit X. It just gives the user an opportunity to say a bit more about why the original commit was problematic.

> $ git revert --no-edit COMMIT_HASH

Author: The problem is that the default text editor is often `vim`, and people does not know how to get out of it 😆

> $ git revert --no-edit COMMIT_HASH

Reviewer 2: So the editor's being tricky reasoning was definitely a part of why the original decision was to use --no-edit everywhere. It's particularly difficult because different people will get different editors then you have to deal with every case in the class in order to proceed and if people are a bit behind it gets worse etc. etc. Even more tricky to deal with online because a helper can't just wander over and see what's going on.

I'm not against re-examining this decision if we want to dedicate the time to it. We could probably do the config to set nano as the default editor across platforms which makes things a bit easier and more consistent.

> $ git revert --no-edit COMMIT_HASH

Reviewer: How about I open an issue for it? I think we might want to revisit this.

My concerns with avoiding talking about editors are that:

1. Students will complete the intermediate-level course without knowing how to write longer commit messages, which is quite an important thing to know how to do
2. If a student forgets to add the `--no-edit` flag they're going to end up trapped in vim anyway

> $ git revert --no-edit COMMIT_HASH

Author: Issue open https://github.com/ImperialCollegeLondon/intermediate_grad_school_git_course/issues/22

>
{: .challenge}

This is yet another good example of why making separate commits for each change is a
good idea, so they can, potentially, be reversed if needed in the future with no fuss.

### Set aside your work safely with `stash`

It is not rare that, while you are working on some feature, you need to check something
else in another branch. Very often this is the case when you want to try some
contributor's code as part of a pull request review process (see next episodes). You
can commit the work you are doing, but if it is not in a state ready to be committed,
what would you do?

`git stash` is the answer. It lets you put your current, not committed work aside in a
special state, reverting the working directory to the way it was in the last commit.
Then, you can easily switch branches, pull new ones or do whatever you want. Once you
are ready to go back to work, you can recover the stashed work as continue as if
nothing had happen.

The following are the `git stash` commands needed to make this happen:

- Stash the current state of the repository, giving some message to remind yourself what
  was this about. The working directory becomes identical to the last commit.
~~~

> $ git stash save "Some informative message"

Author: Neither did I!

~~~
{: .commands}
- List the stashes available in reverse chronological order (last one stashed goes on
  top).
~~~
$ git stash list
~~~
{: .commands}

> - Extracts the stash with the given number from the list, updating the working directory

Reviewer: ```suggestion
- Extract the stash with the given number from the list, updating the working directory
```

(The previous items are in the imperative)

  with its content.
~~~

> $ git stash pop stash@{NUMBER}

Reviewer: I think we should also at least mention that if you run `git stash pop` without an extra arg it just pops the last stash as that's the usual way of using the command. Ditto below for `git stash apply`.

~~~
{: .commands}

> - Applies the given stash without removing it from the list, so you can apply it to

Reviewer: Let's make this imperative too:
```suggestion
- Apply the given stash without removing it from the list, so you can apply it to
```

stash](https://www.atlassian.com/git/tutorials/saving-changes/git-stash).

> ## Practice stashing
>
> Now try how stashing work with the recipe repository. For example, add some
> ingredients, stash the changes, modify the instructions, stash also that. Then have a
> look at the list of stashes and bring those changes back to the working directory
> using `stash pop` and `stash apply`, and see how the list of stashes changes in either
> case.
>
{: .challenge}

### Incorporate past commits with `rebase`

Rebasing is the process of moving or combining a sequence of commits to a new base
commit. In other words, you take a collection of commits that you have created that
branched off a particular commit and make them appear as if they branched off a
different one.

The most common use case for `git rebase` happens when you are working on your feature
branch (let's say `experiment`) and, in the meantime there have been commits done to the
base branch (for example, `main`). You might want to use in your own work some upstream
changes done by someone else or simply keep the history of the repository linear,
facilitating merging back in the future.

> The command is straight forward:

Reviewer: ```suggestion
The command is straightforward:
```

~~~
$ git rebase NEW_BASE
~~~
{: .commands}
where `NEW_BASE` can be either a commit hash or a branch name we want to use as the new
base.

The following figure illustrates the process where, after rebasing, the two commits of
the feature branch have been recreated after the last commit of the main branch.

![Rebase process with a feature branch being moved to another branch]({{ site.baseurl }}https://wac-cdn.atlassian.com/dam/jcr:4e576671-1b7f-43db-afb5-cf8db8df8e4a/01%20What%20is%20git%20rebase.svg?cdnVersion=636
"Rebase process with a feature branch being moved to another branch"){:class="img-responsive"}

> For a very thorough description about how this process work, read this [article on Git

Reviewer: ```suggestion
For a very thorough description about how this process works, read this [article on Git
```

rebase](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase).

> ## Practice rebasing
>
> We are going to practice rebasing in a simple scenario with the recipe repository.
> We need to do some preparatory work to do this:
>
> - Create a `spicy` branch
> - Add some chillies to the list of ingredients and commit the changes
> - Switch back to the `main` branch
> - Add an final step in the instructions indicating that this should be served cold
> - Go back to the `spicy` branch
>
> If you were to add now instructions to chop the chillies finely and put some on top
> of the mix, chances are that you will have conflicts later on when merging back to
> main. We can merge `main` into `spicy`, as we did in the previous episode, but that
> will result in a non-linear history (not a big deal in this case, but things can get
> really complicated).
>

> > So let's use `git rebase` to bring the `spice` branch as it it would have been

Reviewer: ```suggestion
> So let's use `git rebase` to bring the `spicy` branch as it it would have been
```

- "Rewriting history can have unexpected consequences and you risk losing information permanently"
- "Reset: You have made a mistake and want to keep the commit history tidy for the benefit of collaborators"
- "Revert: You want to undo something done in the past without messing too much with the timeline, upsetting your collaborators"
- "Stash: You want to do something else -- e.g. checkout someone else's branch -- without losing your current work"
- "Rebase: Someone else has updated the main branch while you've been working and need to bring those changes to your branch"
---

## Rewriting history with Git

While version control is useful to keep track of changes made to a piece of work over
time, it also let you to modify the timeline of commits. There are several totally
legitimate reasons why you might want to do that, from keeping the commit history clean
of unsuccessful attempts to do something to incorporate work done by someone else.

This episode explores some of the commands `git` offers to manipulate the commit history
for your benefit and that of your collaborators.

### Amend

This is the simplest method of rewriting history: it lets you amend the last commit you

> made, maybe adding some files you forgot to stag or fixing a typo in the commit message.

Reviewer: ```suggestion
made, maybe adding some files you forgot to stage or fixing a typo in the commit message.
```

> After you have made those last minute changes - and `stagged` them, if needed - all you

Reviewer: ```suggestion
After you have made those last minute changes - and `staged` them, if needed - all you
```

While version control is useful to keep track of changes made to a piece of work over
time, it also let you to modify the timeline of commits. There are several totally
legitimate reasons why you might want to do that, from keeping the commit history clean
of unsuccessful attempts to do something to incorporate work done by someone else.

This episode explores some of the commands `git` offers to manipulate the commit history
for your benefit and that of your collaborators.

### Amend

This is the simplest method of rewriting history: it lets you amend the last commit you
made, maybe adding some files you forgot to stag or fixing a typo in the commit message.

After you have made those last minute changes - and `stagged` them, if needed - all you
need to do to amend the last commit while keeping the same commit message is:

```bash
$ git commit --amend --no-edit
```

Or this:

```bash
$ git commit --amend -m "New commit message"
```

> if you want to write a new commit message:

Reviewer: Was this left in by mistake?

Author: yep!

{: .challenge}
> ## Changing History Can Have Unexpected Consequences
>
> Like with `git commit --amend`, using `git reset` to remove a commit is a bad idea if

> > you have **shared** it yet with other people. If you make a commit and share it on

Reviewer: ```suggestion
> you have **already shared** it with other people. If you make a commit and share it on
```

This is yet another good example of why making separate commits for each change is a
good idea, so they can, potentially, be reversed if needed in the future with no fuss.

> > ## `reset` vs `reverse`

Reviewer: Ha, I didn't know `reverse` was a thing!

Reviewer 3: I don't think `reverse` is a command. 

```suggestion
> ## `reset` vs `revert`
```

>
> Both commands let you undo things done in the past, but they both have very different
> use cases.
> - `reset` uses brute force, potentially with destructive consequences, to
> make those changes and is suitable only if the work has not been shared with others
> already. Use when you want to get rid of recent work you're not happy with and start
> all over again.

> > - `reverse` is more light weight and surgical, to target specific changes and

Reviewer: ```suggestion
> - `reverse` is more lightweight and surgical, to target specific changes and
```

Reviewer 3: ```suggestion
> - `revert` is more lightweight and surgical, to target specific changes and
```

can commit the work you are doing, but if it is not in a state ready to be committed,
what would you do?

`git stash` is the answer. It lets you put your current, uncommitted work aside in a

> special state, turning the working directory to the way it was in the last commit.

Reviewer: ```suggestion
special state, turning the working directory back to the way it was in the last commit.
```
